network/p2p/p2pbuilder/libp2pNodeBuilder.go

```diff
@@ -391,20 +365,31 @@ func (builder *LibP2PNodeBuilder) Build() (p2p.LibP2PNode, error) {
 
 	cm := component.NewComponentManagerBuilder().
 		AddWorker(func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+			// routing system is created here, because it needs to be created during the node startup.
 			rsys, err := builder.buildRouting(ctx, h)
 			if err != nil {
-				ctx.Throw(fmt.Errorf("could not create routing system: %w", err))
+				ctx.Throw(fmt.Errorf("could not create rsys system: %w", err))
 			}
 			node.SetRouting(rsys)
+			builder.gossipSubBuilder.SetRoutingSystem(rsys)
 
-			gossipSub, err := builder.buildGossipSub(ctx, rsys, h)
+			// gossipsub is created here, because it needs to be created during the node startup.
+			gossipSub, scoreTracer, err := builder.gossipSubBuilder.Build(ctx)
 			if err != nil {
 				ctx.Throw(fmt.Errorf("could not create gossipsub: %w", err))
 			}
+			if scoreTracer != nil {
+				node.SetPeerScoreExposer(scoreTracer)
+			}
 			node.SetPubSub(gossipSub)
 
 			ready()
+		}).
+		AddWorker(func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+			// encapsulates shutdown logic for the libp2p node.
```

Comment:
I'm curious why you split this into a second worker. It's functionally equivalent, but a little confusing. What would happen in `node.Stop()` was called before completing the setup in the first worker?

Reply:
It was to address this comment: https://github.com/onflow/flow-go/pull/3986#discussion_r1137336133, however, I think splitting the startup of GossipSub in the first worker and the libp2p node shutdown in the second worker yields a cleaner and more maintainable approach as it does a separation of concerns pattern (at the cost of an extra goroutine, which I think is negligible).